keystore/v2/keystore/filesystem/export.go

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filesystem
+
+import (
+	"errors"
+	"time"
+
+	"github.com/cossacklabs/acra/keystore/v2/keystore/api"
+	"github.com/cossacklabs/acra/keystore/v2/keystore/asn1"
+	"github.com/cossacklabs/acra/keystore/v2/keystore/crypto"
+	backendAPI "github.com/cossacklabs/acra/keystore/v2/keystore/filesystem/backend/api"
+	"github.com/cossacklabs/acra/keystore/v2/keystore/signature"
+	"github.com/cossacklabs/acra/utils"
+)
+
+// Errors returned by export/import routines.
+var (
+	ErrKeyRingExists = errors.New("imported key ring already exists")
+)
+
+func (s *KeyStore) exportKeyRings(paths []string) (rings []asn1.KeyRing, err error) {
+	rings = make([]asn1.KeyRing, len(paths))
+	defer func() {
+		if err != nil {
+			zeroizeKeyRings(rings)
+		}
+	}()
+	for i, path := range paths {
+		err := s.exportKeyRing(path, &rings[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+	return rings, nil
+}
+
+func (s *KeyStore) exportKeyRing(path string, ringData *asn1.KeyRing) error {
+	ring := newKeyRing(s, path)
+	err := s.readKeyRing(ring)
+	if err != nil {
+		return err
+	}
+	*ringData, err = ring.exportASN1()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *KeyStore) importKeyRing(newRingData *asn1.KeyRing, delegate *api.KeyRingImportDelegate) error {
+	keyRing := newKeyRing(s, string(newRingData.Purpose))
+	err := s.readKeyRing(keyRing)
+	switch err {
+	case nil:
+		// If the key store successfuly returned an existing key ring with the same name,
+		// we have to resolve this conflict somehow. Present both current and new versions
+		// to the delegate and let it decide how to proceed.
+		decision := api.ImportAbort
+		err := ErrKeyRingExists
+		if delegate != nil && delegate.DecideKeyRingOverwrite != nil {
```

Comment:
that is why struct seems like not a comfortable choice in use. imho better to provide interface for strategy pattern which will be used on importing when keyring already exists. And provide several types of strategy: "skip"/"overwrite"/"abort". And here just call `delegate.ImportExistingData` or something like that. Which delegate to use will be chosen on start by keystore configuration and some fabric method like:
```
switch importStrategy {
case api.ImportOverwrite:
  return NewImportOverwriteStrategy()
case api.ImportSkip:
  return NewImportSkipStrategy()
case api.AbortOnOverwrite:
  return NewAbortStrategy()
}
```

then here will be more clean code, one switch branch on start and zero in runtime execution here.

Reply:
> And here just call delegate.ImportExistingData or something like that. 

That would look a bit neater from our side but that will couple the delegate implementation to the key store implementation, without a particularly good reason. If we use enumeration values then the delegate can be reused with different key store implementations without any issues.